### DISCUSSION TITLE
Properly handle negative volume meshes

### DISCUSSION THREAD

**Discussion by pca006132** [elalish/manifold#693] [General]
```js
const {cube, sphere} = Manifold;
const foo = cube(3, true).subtract(cube(1, true));
const parts = foo.decompose();
console.log(parts.length);
const result = parts[1];
```

The result has infinite size (that we do not compute correctly), while not violating the manifoldness definition in https://github.com/elalish/manifold/wiki/Manifold-Library#manifoldness-definition I am not sure if this is expected.

**Comment by elalish** [comment]
The output I get from your script is:
```
2
Triangles: 12
Bounding Box: X = 1 mm, Y = 1 mm, Z = 1 mm
Genus: 0, Volume: -0 cm^3
Took 0.02 seconds
```
What is infinite? As far as I can tell it's just saying you have a negative volume cube, which is correct.

**Reply by pca006132** [reply]
My understanding of the cube is that it contains everything except the 1mm^3 volume, so the bounding box and volume should be infinite.

**Reply by elalish** [reply]
I see what you mean from a theoretical perspective, but I choose a more computationally practical approach of saying this inverted cube has the same bounding box (its verts) and a negative volume. I think that's a more common way to think about it, for both polygons and polyhedra.

**Reply by pca006132** [reply]
Yeah I get what you mean. I just wasn't sure about whether our boolean ops work on these negative volumes (e.g. positive union negative is equivalent to difference of positive and positive).

Also, not sure if it would be a nice idea to relate these negative volumes to their bounding volume for decompose output.

**Reply by elalish** [reply]
I think they work - and I think they are consistent with Clipper's boolean in 2D. If we find any cases where they are not consistent, or otherwise confusing, let's discuss. But I don't think I want to allow the concept of unbounded solids into our floating-point library.

**Reply by pca006132** [reply]
I don't think it is working properly (at least not in the way I would expect):
```js
const {cube, sphere} = Manifold;
const foo = cube(10, true).subtract(cube(5, true));
const parts = foo.decompose();
const result = cube(10).add(parts[1]);
console.log(result.decompose().length);
```
![image](https://github.com/elalish/manifold/assets/12198657/0e084cf1-b867-42f8-83cc-943f5abed54d)

If we do subtract instead of add, result is just a `cube(10)` and the decomposition result still has length 2.

**Reply by elalish** [reply]
Huh, that's interesting. I'm going to have to think about that...

**Reply by elalish** [reply]
We may be able to shift the winding number calculation by 1 if the volume is negative. I'm curious if that'll cause other problems...

**Comment by pca006132** [comment]
btw should we convert this to an issue?

**Reply by elalish** [reply]
I dunno - I think the main problem is just that negative volumes are a bit counter-intuitive. And they render oddly (due to back-face culling, which becomes front-face culling). I still think it would be worth some some comparisons between 2D polygon booleans and 3D booleans. They should be consistent, but intuitive may be a bridge too far.

**Reply by pca006132** [reply]
what about the issue of boolean operation with negative volume? it does not work now (see above)

**Reply by elalish** [reply]
That's the thing - I'm not sure it's "not working" now. Can you make the equivalent polygons (negative area) and check what Clipper's Booleans do with them in 2D? I think it may be the same.

**Reply by pca006132** [reply]
I cannot make a negative area polygon in Clipper, our constructor calls union and removes the negative part. Also, it seems that the union algorithm in clipper correctly handles negative contour that overlaps with positive contour:

```js
const c = CrossSection.ofPolygons([[[0, 0], [1, 0], [0, 1]], [[0.5, 0], [0.5, 1], [1.5, 0]]])
const result = c.extrude(1)
```

![image](https://github.com/elalish/manifold/assets/12198657/46adb6cb-f487-4ca3-be9a-a1feadb232e7)

**Reply by elalish** [reply]
Fair enough, see #745
